File: jvm/diffuser/src/test/java/com/spotify/diffuser/fuser/FuserLaws.kt
/*
 * Copyright (c) 2019 Spotify AB.
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.spotify.diffuser.fuser

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Shared tests that verifies the laws that all Fusers are expected to conform to.
 */
abstract class FuserLaws {

    abstract fun identityFuser(source: Source<Int>): Fuser<Int>

    @Test
    fun `Values are dispatched in order`() {
        // events after connect are sent in order
        val source = TestSource()
        val fuser = identityFuser(source)
        val output = mutableListOf<Int>()

        val disposable = fuser.connect { output.add(it) }
        source.emit(1)
        source.emit(2)
        source.emit(3)
        disposable.dispose()

        assertEquals(listOf(1, 2, 3), output)
    }

    @Test
    fun `Values are only dispatched after connect()`() {
        // events after connect are dispatched
        val source = TestSource()
        val fuser = identityFuser(source)
        val output = mutableListOf<Int>()

        source.emit(1)
        val disposable = fuser.connect { output.add(it) }
        source.emit(2)
        source.emit(3)
        disposable.dispose()

        assertEquals(listOf(2, 3), output)
    }

    @Test
    fun `Values are not dispatched after dispose()`() {
        val source = AfterDisposeEmittingSource()
        val fuser = identityFuser(source)
        val output = mutableListOf<Int>()

        val disposable = fuser.connect { output.add(it) }
        source.emit(1)
        source.emit(2)
        disposable.dispose()
        source.emit(3)

        assertEquals(listOf(1, 2), output)
    }

    @Test
    fun `Calling dispose() disposes the source`() {
        val source = DisposableSource()
        val fuser = identityFuser(source)

        val disposable = fuser.connect {}

        assertFalse(source.isDisposed)
        disposable.dispose()
        assertTrue(source.isDisposed)
    }

    @Test
    fun `Can be connected to multiple times`() {
        val source = TestSource()
        val fuser = identityFuser(source)

        val output1 = mutableListOf<Int>()
        val connection1 = fuser.connect { output1.add(it) }

        val output2 = mutableListOf<Int>()
        val connection2 = fuser.connect { output2.add(it) }

        val input = listOf(1, 2, 3)
        input.forEach { source.emit(it) }

        connection1.dispose()
        connection2.dispose()

        assertEquals(input, output1)
        assertEquals(input, output2)
    }

    @Test
    fun `Connections can be disposed independently`() {
        val source = TestSource()
        val fuser = identityFuser(source)

        val output1 = mutableListOf<Int>()
        val connection1 = fuser.connect { output1.add(it) }

        val output2 = mutableListOf<Int>()
        val connection2 = fuser.connect { output2.add(it) }

        source.emit(1)
        source.emit(2)
        connection1.dispose()
        source.emit(3)
        connection2.dispose()

        assertEquals(listOf(1, 2), output1)
        assertEquals(listOf(1, 2, 3), output2)
    }
}

File: swift/Diffuser/Tests/DiffuserTests/fuser/FuserLaws.swift
// Copyright (c) 2019 Spotify AB.
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import Diffuser
import SwiftCheck
import XCTest

class FuserLaws {

    private let identityFuser: (AnySource<Int>) -> Fuser<Int>

    init(fuserFrom fuserFactory: @escaping (AnySource<Int>) -> Fuser<Int>) {
        self.identityFuser = fuserFactory
    }

    func verify() {
        valuesAreDispatchedInOrder()
        valuesAreOnlyDispatchedAfterConnect()
        valuesAreNotDispatchedAfterDispose()
        callingDisposeDisposesTheSource()
        canBeConnectedToMultipleTimes()
        connectionsCanBeDisposedIndependently()
    }

    func valuesAreDispatchedInOrder() {
        // events after connect are sent in order
        let source = TestSource()
        let fuser = identityFuser(AnySource(source))
        var output: [Int] = []

        let disposable = fuser.connect { output.append($0) }
        source.emit(1)
        source.emit(2)
        source.emit(3)
        disposable.dispose()

        XCTAssertEqual([1, 2, 3], output)
    }

    func valuesAreOnlyDispatchedAfterConnect() {
        // events after connect are dispatched
        let source = TestSource()
        let fuser = identityFuser(AnySource(source))
        var output: [Int] = []

        source.emit(1)
        let disposable = fuser.connect { output.append($0) }
        source.emit(2)
        source.emit(3)
        disposable.dispose()

        XCTAssertEqual([2, 3], output)
    }

    func valuesAreNotDispatchedAfterDispose() {
        let source = AfterDisposeEmittingSource()
        let fuser = identityFuser(AnySource(source))
        var output: [Int] = []

        let disposable = fuser.connect { output.append($0) }
        source.emit(1)
        source.emit(2)
        disposable.dispose()
        source.emit(3)

        XCTAssertEqual([1, 2], output)
    }

    func callingDisposeDisposesTheSource() {
        let source = DisposableSource()
        let fuser = identityFuser(AnySource(source))

        let disposable = fuser.connect { _ in }

        XCTAssertFalse(source.isDisposed)
        disposable.dispose()
        XCTAssertTrue(source.isDisposed)
    }

    func canBeConnectedToMultipleTimes() {
        let source = TestSource()
        let fuser = identityFuser(AnySource(source))

        var output1: [Int] = []
        let connection1 = fuser.connect { output1.append($0) }

        var output2: [Int] = []
        let connection2 = fuser.connect { output2.append($0) }

        let input = [1, 2, 3]
        input.forEach { source.emit($0) }

        connection1.dispose()
        connection2.dispose()

        XCTAssertEqual(input, output1)
        XCTAssertEqual(input, output2)
    }

    func connectionsCanBeDisposedIndependently() {
        let source = TestSource()
        let fuser = identityFuser(AnySource(source))

        var output1: [Int] = []
        let connection1 = fuser.connect { output1.append($0) }

        var output2: [Int] = []
        let connection2 = fuser.connect { output2.append($0) }

        source.emit(1)
        source.emit(2)
        connection1.dispose()
        source.emit(3)
        connection2.dispose()

        XCTAssertEqual([1, 2], output1)
        XCTAssertEqual([1, 2, 3], output2)
    }
}
